### ISSUE TITLE
How is the dev set used in MZ during training?

### ISSUE THREAD

**Issue by avoidthevoid** [NTMC-Community/MatchZoo#399]
Hi,

I am using MZ to rank a dataset different from the examples. Below I have put a snippet of the logs created by the ARCII model in the training phase. It shows 5 iterations that each consist of three steps (Train:train, Eval:test, Eval:valid). Now I know the first step is just training the model, but then I am wondering:
-How are **Eval:test** and **Eval:valid** used in the model during the training phase? Or is only the loss in the training set used to improve the model?
-Does MatchZoo also train on the **dev set** or do you have to specify this when calling the model for the test phase (so that it predicts on the dev set instead of the test set)?

```
[Train:train]	Iter:0	loss=2.298408		
[Eval:test]	Iter:0	ndcg@3=0.041360	map=0.094886	ndcg@5=0.046609
[Eval:valid]	Iter:0	ndcg@3=0.070182	map=0.129720	ndcg@5=0.091899
[Train:train]	Iter:1	loss=1.526796		
[Eval:test]	Iter:1	ndcg@3=0.032534	map=0.089728	ndcg@5=0.036889
[Eval:valid]	Iter:1	ndcg@3=0.057909	map=0.119551	ndcg@5=0.083259
[Train:train]	Iter:2	loss=1.169270		
[Eval:test]	Iter:2	ndcg@3=0.036380	map=0.087933	ndcg@5=0.038502
[Eval:valid]	Iter:2	ndcg@3=0.058517	map=0.113737	ndcg@5=0.087344
[Train:train]	Iter:3	loss=1.048904		
[Eval:test]	Iter:3	ndcg@3=0.032265	map=0.089057	ndcg@5=0.037902
[Eval:valid]	Iter:3	ndcg@3=0.066886	map=0.121801	ndcg@5=0.094897
[Train:train]	Iter:4	loss=1.037611		
[Eval:test]	Iter:4	ndcg@3=0.036799	map=0.092546	ndcg@5=0.042928
[Eval:valid]	Iter:4	ndcg@3=0.074272	map=0.126365	ndcg@5=0.104420
[Train:train]	Iter:5	loss=1.041891		
[Eval:test]	Iter:5	ndcg@3=0.032534	map=0.091794	ndcg@5=0.036747
[Eval:valid]	Iter:5	ndcg@3=0.076984	map=0.129180	ndcg@5=0.107115
```

Look forward to your reply, thanks in advance!

**Comment by faneshion** [comment]
You are right, only the loss of training set is used to improve the model. We do not train the model on dev set. You can choose to config the Eval:valid part or not. The valid set is used to choose the final model. The Eval:test set is to show how the performance varies along with Eval:valid set. While in predicting, it is no need to config the valid part.

**Comment by avoidthevoid** [comment]
I see, thank you for your reply!

**Comment by avoidthevoid** [comment]
Hi!

I was still wondering, about one thing: how does MatchZoo select which (best) weights to use in the final prediction? Is this based on the performance of the model (using those weights) on the dev set as calculated by (a mean? of) the metrics as configured in the config file (e.g. MAP, NDCG@3)?

Thanks in advance!

**Comment by DaoD** [comment]
@avoidthevoid Hi, I think the model doesn't select weights automatically. You need to save the weights manually and compare the loss on the dev set to choose which model to use.
In keras, you can use the checkpoint to save the model in a fixed period.

Fro example:
callback = ModelCheckpoint(filepath, monitor='val_loss', verbose=0, save_best_only=True, save_weights_only=True, mode='auto', period=1)
model.fit_generator(train_generator, epochs=20, callbacks=[callback])

After the training process, you can use model.backend.load_weights("xxx", by_name=True) to load the best model.
